aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -80,5 +60,49 @@ Tensor _segment_reduce_cpu(
   return output;
 }
 
+} // namespace
+
+enum SegmentReductionType { MAX };
+static const std::map<std::string, SegmentReductionType> segmentReduce2REDUCE =
+    {
+        {"max", MAX},
+};
+
+Tensor _segment_reduce_(
+    const Tensor& data,
+    std::string reduce,
+    const c10::optional<Tensor>& lengths,
+    const c10::optional<Tensor>& indices,
+    int64_t axis,
+    bool unsafe) {
+  axis = maybe_wrap_dim(axis, data.ndimension());
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
+  TORCH_CHECK(data.dim() == 1);
+  TORCH_CHECK(data.numel() > 0);
+  TORCH_CHECK(
+      at::native::segmentReduce2REDUCE.at(reduce) == MAX,
+      "Currently only 'max' reduction is supported!");
+
+  // length related checks
+  TORCH_CHECK(
+      lengths.has_value() && !indices.has_value(),
```

Comment:
Do you have a follow-up to support indices-based reduction? That would allow you to not compute cumsum

Reply:
next steps are roughly in order are:
- support more reduction types
- add backward support
- add more documentation/testing/benchmark
- add support for multi dimension input
- add support for indices based reduction

So it might take a while to come back to indices since not very high priority.